### PR TITLE
[Quest API] Add EVENT_DAMAGE_GIVEN and EVENT_DAMAGE_TAKEN to Perl/Lua.

### PR DESCRIPTION
**Pull request by Kinglykrab**
# Perl
- Add `EVENT_DAMAGE_GIVEN`.
- Add `EVENT_DAMAGE_TAKEN`.
- Both events export `$entity_id`, `$damage`, `$spell_id`, `$skill_id`, `$is_damage_shield`, `$is_avoidable`, `$buff_slot`, `$is_buff_tic`, `$special_attack`.

# Lua
- Add `event_damage_given`.
- Add `event_damage_taken`.
- Both events export `e.entity_id`, `e.damage`, `e.spell_id`, `e.skill_id`, `e.is_damage_shield`, `e.is_avoidable`, `e.buff_slot`, `e.is_buff_tic`, `e.special_attack`, and `e.other`.

# Notes
- These events allow operators to have events fire based on damage given or taken, as well as keep better track of a Bot, Client, or NPC's damage per second or otherwise.
- Special Attack is only useful for NPCs, but allows you to see if the attack is Rampage, AERampage, or Chaotic Stab.

### CODE REVIEW

File: zone/lua_parser_events.cpp [hunk 1]
```diff
@@ -453,6 +453,50 @@ void handle_npc_despawn_zone(
 	lua_setfield(L, -2, "other");
 }
 
+void handle_npc_damage(
+	QuestInterface *parse,
+	lua_State* L,
+	NPC* npc,
+	Mob* init,
+	std::string data,
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:461]: `const std::string& data,`

File: zone/lua_parser_events.cpp [hunk 2]
```diff
@@ -453,6 +453,50 @@ void handle_npc_despawn_zone(
...
+	lua_pushnumber(L, std::stoll(sep.arg[1]));
+	lua_setfield(L, -2, "damage");
+
+	lua_pushnumber(L, std::stoi(sep.arg[2]));
+	lua_setfield(L, -2, "spell_id");
+
+	lua_pushnumber(L, std::stoi(sep.arg[3]));
+	lua_setfield(L, -2, "skill_id");
+
+	lua_pushboolean(L, std::stoi(sep.arg[4]) == 0 ? false : true);
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:479]: Simplify: `lua_pushboolean(L, std::stoi(sep.arg[4]) != 0);`

File: zone/lua_parser_events.cpp [hunk 3]
```diff
@@ -453,6 +453,50 @@ void handle_npc_despawn_zone(
...
+	lua_setfield(L, -2, "is_damage_shield");
+
+	lua_pushboolean(L, std::stoi(sep.arg[5]) == 0 ? false : true);
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:482]: Simplify: lua_pushboolean(L, std::stoi(sep.arg[5]) != 0);

File: zone/lua_parser_events.cpp [hunk 4]
```diff
@@ -453,6 +453,50 @@ void handle_npc_despawn_zone(
...
+	lua_setfield(L, -2, "is_avoidable");
+
+	lua_pushnumber(L, std::stoi(sep.arg[6]));
+	lua_setfield(L, -2, "buff_slot");
+
+	lua_pushboolean(L, std::stoi(sep.arg[7]) == 0 ? false : true);
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:488]: Simplify: lua_pushboolean(L, std::stoi(sep.arg[7]) != 0);

File: zone/lua_parser_events.cpp [hunk 5]
```diff
@@ -1168,6 +1212,51 @@ void handle_player_bot_create(
 	lua_setfield(L, -2, "bot_gender");
 }
 
+void handle_player_damage(
+	QuestInterface *parse,
+	lua_State* L,
+	Client* client,
+	std::string data,
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:1219]: const reference.

File: zone/lua_parser_events.cpp [hunk 6]
```diff
@@ -1168,6 +1212,51 @@ void handle_player_bot_create(
...
+	lua_pushnumber(L, std::stoll(sep.arg[1]));
+	lua_setfield(L, -2, "damage");
+
+	lua_pushnumber(L, std::stoi(sep.arg[2]));
+	lua_setfield(L, -2, "spell_id");
+
+	lua_pushnumber(L, std::stoi(sep.arg[3]));
+	lua_setfield(L, -2, "skill_id");
+
+	lua_pushboolean(L, std::stoi(sep.arg[4]) == 0 ? false : true);
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:1237]: same as above.

File: zone/lua_parser_events.cpp [hunk 7]
```diff
@@ -1168,6 +1212,51 @@ void handle_player_bot_create(
...
+	lua_setfield(L, -2, "is_damage_shield");
+
+	lua_pushboolean(L, std::stoi(sep.arg[5]) == 0 ? false : true);
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:1240]: same as above.

File: zone/lua_parser_events.cpp [hunk 8]
```diff
@@ -1168,6 +1212,51 @@ void handle_player_bot_create(
...
+	lua_setfield(L, -2, "is_avoidable");
+
+	lua_pushnumber(L, std::stoi(sep.arg[6]));
+	lua_setfield(L, -2, "buff_slot");
+
+	lua_pushboolean(L, std::stoi(sep.arg[7]) == 0 ? false : true);
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:1246]: same as above.

File: zone/lua_parser_events.cpp [hunk 9]
```diff
@@ -1168,6 +1212,51 @@ void handle_player_bot_create(
...
+	lua_setfield(L, -2, "is_buff_tic");
+
+	lua_pushnumber(L, std::stoi(sep.arg[8]));
+	lua_setfield(L, -2, "special_attack");
+
+	if (extra_pointers && extra_pointers->size() >= 1) {
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:1252]: Use empty() `if (extra_pointers && !extra_pointers->empty()) {`

File: zone/lua_parser_events.cpp [hunk 10]
```diff
@@ -1986,22 +2075,39 @@ void handle_bot_unequip_item(
 	uint32 extra_data,
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:2075]: const reference.

File: zone/lua_parser_events.cpp [hunk 11]
```diff
@@ -1986,22 +2075,39 @@ void handle_bot_unequip_item(
...
-	l_item_o.push(L);
-	lua_setfield(L, -2, "item");
-}
+	lua_pushnumber(L, std::stoi(sep.arg[2]));
+	lua_setfield(L, -2, "spell_id");
+
+	lua_pushnumber(L, std::stoi(sep.arg[3]));
+	lua_setfield(L, -2, "skill_id");
+
+	lua_pushboolean(L, std::stoi(sep.arg[4]) == 0 ? false : true);
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:2092]: same as above.

File: zone/lua_parser_events.cpp [hunk 12]
```diff
@@ -1986,22 +2075,39 @@ void handle_bot_unequip_item(
...
+	lua_setfield(L, -2, "is_damage_shield");
 
+	lua_pushboolean(L, std::stoi(sep.arg[5]) == 0 ? false : true);
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:2095]: same as above.

File: zone/lua_parser_events.cpp [hunk 13]
```diff
@@ -1986,22 +2075,39 @@ void handle_bot_unequip_item(
...
+	lua_setfield(L, -2, "is_avoidable");
+
+	lua_pushnumber(L, std::stoi(sep.arg[6]));
+	lua_setfield(L, -2, "buff_slot");
+
+	lua_pushboolean(L, std::stoi(sep.arg[7]) == 0 ? false : true);
```
Comment (Aeadoin) [inline zone/lua_parser_events.cpp:2101]: same as above.

File: zone/lua_parser_events.h [hunk 14]
```diff
@@ -210,6 +210,16 @@ void handle_npc_despawn_zone(
 	std::vector<std::any> *extra_pointers
 );
 
+void handle_npc_damage(
+	QuestInterface *parse,
+	lua_State* L,
+	NPC* npc,
+	Mob *init,
+	std::string data,
```
Comment (Aeadoin) [inline zone/lua_parser_events.h:218]: const reference.

File: zone/lua_parser_events.h [hunk 15]
```diff
@@ -652,6 +662,14 @@ void handle_player_augment_remove(
 	std::vector<std::any> *extra_pointers
 );
 
+void handle_player_damage(
+	QuestInterface *parse,
+	lua_State* L,
+	Client* client,
+	std::string data,
```
Comment (Aeadoin) [inline zone/lua_parser_events.h:669]: const reference.

**Review by Aeadoin — COMMENTED** [review]
Overall looks good to me, just some minor smell.

File: zone/attack.cpp [hunk 16]
```diff
@@ -3820,6 +3820,98 @@ void Mob::CommonDamage(Mob* attacker, int64 &damage, const uint16 spell_id, cons
...
+				avoidable ? 1 : 0,
+				buffslot,
+				iBuffTic ? 1 : 0,
+				static_cast<int>(special)
+			);
+
+			if (IsBot() && has_bot_taken_event) {
+				parse->EventBot(EVENT_DAMAGE_TAKEN, CastToBot(), attacker ? attacker : nullptr, export_string, 0);
+			} else if (IsClient() && has_player_taken_event) {
+				args.push_back(attacker ? attacker : nullptr);
```
[the remark this answers is not in the document]
Comment (Kinglykrab) [reply zone/attack.cpp:3884]: No where else in the source does that.

File: zone/lua_parser_events.cpp (hunk 1 above)
Comment (Kinglykrab) [reply zone/lua_parser_events.cpp:461]: Every other event doesn't do this.

File: zone/lua_parser_events.cpp (hunk 3 above)
Comment (Kinglykrab) [reply zone/lua_parser_events.cpp:482]: Same thing as what I have.

File: zone/lua_parser_events.cpp (hunk 9 above)
Comment (Kinglykrab) [reply zone/lua_parser_events.cpp:1252]: Same thing as what I'm doing here.

**Review by Aeadoin — APPROVED** [review]
looks good, items pointed out are nitpicks... can be cleaned up if someone ever wants to take a stab at cleaning up lua_parser_events.cpp

File: zone/attack.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (Akkadius) [reply zone/attack.cpp:3884]: We use emplace_back in tons of places

[the remark this answers is not in the document]
Comment (Kinglykrab) [reply zone/attack.cpp:3884]: I just mean every spot we use `args` we use `push_back`.

**Pull request merged**